src/db.rs

```diff
@@ -85,8 +83,73 @@ pub fn insert_connection(connection: Connection) -> Result<Uuid> {
     Ok(nwuuid)
 }
 
+pub fn insert_policy(policy: &policy::Policy) -> Result<u64> {
+    let mut dbconn = get_dbconn()?;
+
+    let allowed_digests_json = serde_json::to_string(&policy.allowed_digests)?;
+    let allowed_policy_json = serde_json::to_string(&policy.allowed_policies)?;
+    let allowed_build_ids_json = serde_json::to_string(&policy.allowed_build_ids)?;
+
+    let mysqlstr = format!(
+        "INSERT INTO policy (allowed_digests, allowed_policies, min_fw_api_major, min_fw_api_minor,
+                               allowed_build_ids, create_date, valid)
+                               VALUES({:?}, {:?}, {}, {}, {:?}, NOW(), 1)",
+        allowed_digests_json,
+        allowed_policy_json,
+        policy.min_fw_api_major,
+        policy.min_fw_api_minor,
+        allowed_build_ids_json
+    );
+    dbconn.query::<u64, String>(mysqlstr)?;
+    Ok(dbconn.last_insert_id())
+}
+
+pub fn get_policy(pid: u64) -> Result<policy::Policy> {
+    let mut dbconn = get_dbconn()?;
+
+    let pidstr = pid.to_string();
+    let mqstr = "SELECT allowed_digests, allowed_policies, min_fw_api_major, min_fw_api_minor, allowed_build_ids, create_date, delete_date  FROM policy WHERE id = ? AND valid = 1";
+
+    let qrow: mysql::Row = dbconn.exec_first(mqstr, (pidstr,))?.unwrap();
+
+    let ret_pol: policy::Policy = {
+        policy::Policy {
+            allowed_digests: serde_json::from_str(&mysql::from_value_opt::<String>(
+                qrow["allowed_digests"].clone(),
+            )?)
+            .unwrap(),
+            allowed_policies: serde_json::from_str(&mysql::from_value_opt::<String>(
+                qrow["allowed_policies"].clone(),
+            )?)
+            .unwrap(),
+            min_fw_api_major: mysql::from_value_opt::<i64>(qrow["min_fw_api_major"].clone())?
+                .try_into()
+                .unwrap(),
+            min_fw_api_minor: mysql::from_value_opt::<i64>(qrow["min_fw_api_minor"].clone())?
+                .try_into()
+                .unwrap(),
+            allowed_build_ids: serde_json::from_str(&mysql::from_value_opt::<String>(
+                qrow["allowed_build_ids"].clone(),
+            )?)
+            .unwrap(),
```

Comment:
That's a lot of code for 5 fields. Never did this, but the example in the mysql crate README is:

```rust
#[derive(Debug, PartialEq, Eq)]
struct Payment {
    customer_id: i32,
    amount: i32,
    account_name: Option<String>,
}

...

// Let's select payments from database. Type inference should do the trick here.
let selected_payments = conn
    .query_map(
        "SELECT customer_id, amount, account_name from payment",
        |(customer_id, amount, account_name)| {
            Payment { customer_id, amount, account_name }
        },
    )?;
```

Looks like this should work for us as well (we just need to add some serde_json::from_str() to the mix.

(IMO this can wait for a later fixup.)

Reply:
I agree.  I was trying to use the built in types for mysql::Row, but should have just followed the pattern from get_secret_policy.  I will fix this later as you suggest.